File: clients/android/AwareChat-Android/app/src/main/java/com/example/awarechat_android/core/network/ApiClient.kt
package com.example.awarechat_android.core.network

import com.example.awarechat_android.core.protocol.HealthResponse
import com.example.awarechat_android.core.protocol.UserDto
import com.example.awarechat_android.core.protocol.UsersResponse

interface ApiClientContract {
    suspend fun health(): HealthResponse
    suspend fun users(): List<UserDto>
}

class ApiClient(
    private val configuration: NetworkConfiguration,
    private val networkClient: NetworkClient = NetworkManager(),
) : ApiClientContract {
    override suspend fun health(): HealthResponse = networkClient.fetch(
        request = ApiEndpoint.Health.request(configuration),
        deserializer = HealthResponse.serializer(),
    )

    override suspend fun users(): List<UserDto> = networkClient.fetch(
        request = ApiEndpoint.Users.request(configuration),
        deserializer = UsersResponse.serializer(),
    ).users
}

File: clients/ios/AwareChat-iOS/AwareChat-iOS/Core/Networking/APIClient.swift
import Foundation

nonisolated protocol APIClientProtocol: Sendable {
  func health() async throws -> HealthResponse
  func users() async throws -> [UserDTO]
}

nonisolated final class APIClient: APIClientProtocol, Sendable {
  private let configuration: NetworkConfiguration
  private let networkManager: any NetworkProtocol

  init(
    configuration: NetworkConfiguration,
    networkManager: any NetworkProtocol = NetworkManager()
  ) {
    self.configuration = configuration
    self.networkManager = networkManager
  }

  func health() async throws -> HealthResponse {
    try await networkManager.fetch(
      .health,
      configuration: configuration,
      as: HealthResponse.self
    )
  }

  func users() async throws -> [UserDTO] {
    let response = try await networkManager.fetch(
      .users,
      configuration: configuration,
      as: UsersResponse.self
    )
    return response.users
  }
}
